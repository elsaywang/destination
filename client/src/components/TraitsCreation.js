import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Add from '@react/react-spectrum/Icon/Add';
import Button from '@react/react-spectrum/Button';
import Link from '@react/react-spectrum/Link';
import styles from './TraitsCreation.css';

class TraitsCreation extends Component {
    singleRowTraitsCreation = () => {
        const { traitsCreationLabelText } = this.props;
        return (
            <Link href="#">
                <div className={styles.singleCreation}>
                    <Add size="S" />
                    <div className={styles.singleCreationlinkText}>{traitsCreationLabelText}</div>
                </div>
            </Link>
        );
    };

    multiRowsTraitsCreation = () => {
        const { handleTraitsCreation, selectedSignals, traitsCreationLabelText } = this.props;

        const { selectionMessage, warning } = selectedSignals;

        return (
            <div className={styles.multiCreation}>
                <span className={styles.message}>{selectionMessage}</span>
                <Button
                    label={traitsCreationLabelText}
                    icon={<Add />}
                    onClick={handleTraitsCreation}
                    variant="action"
                    disabled={warning}
                />
            </div>
        );
    };

    render() {
        const { multiCreation, handleTraitsCreation } = this.props;
        return multiCreation ? this.multiRowsTraitsCreation() : this.singleRowTraitsCreation();
    }
}

TraitsCreation.propTypes = {
    multiCreation: PropTypes.bool,
    selectedSignals: PropTypes.object,
    traitsCreationLabelText: PropTypes.string,
    handleTraitsCreation: PropTypes.func,
};

export default TraitsCreation;
