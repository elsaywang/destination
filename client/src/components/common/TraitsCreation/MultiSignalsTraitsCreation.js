import React from 'react';
import PropTypes from 'prop-types';
import styles from './TraitsCreation.css';
import Button from '@react/react-spectrum/Button';
import Add from '@react/react-spectrum/Icon/Add';

const MultiSignalsTraitsCreation = ({
    selectedSignals,
    storeSessionAndNavigateToTraits,
    isMaxSignalSelectionsReached,
}) => {
    const {
        selectionMessage,
        hasSignalSelectionsTypeWarning,
        hasOnboardedSignalSelectionsWarning,
    } = selectedSignals;

    const isTraitCreationDisabled =
        hasSignalSelectionsTypeWarning ||
        isMaxSignalSelectionsReached ||
        hasOnboardedSignalSelectionsWarning;

    return (
        <div className={styles.multiCreation}>
            <span className={styles.message}>{selectionMessage}</span>
            <Button
                onClick={storeSessionAndNavigateToTraits}
                label="Create Trait From Multiple Signals"
                icon={<Add />}
                variant="action"
                disabled={isTraitCreationDisabled}
                data-test="multi-signals-trait-creation"
            />
        </div>
    );
};

MultiSignalsTraitsCreation.propTypes = {
    selectedSignals: PropTypes.shape({
        selectionMessage: PropTypes.string,
        hasSignalSelectionsTypeWarning: PropTypes.bool,
        hasOnboardedSignalSelectionsWarning: PropTypes.bool,
        selectedRowIndexes: PropTypes.array,
    }),
    storeSessionAndNavigateToTraits: PropTypes.func.isRequired,
    isMaxSignalSelectionsReached: PropTypes.bool,
};

export default MultiSignalsTraitsCreation;
