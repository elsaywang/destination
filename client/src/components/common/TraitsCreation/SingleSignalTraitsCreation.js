import React from 'react';
import PropTypes from 'prop-types';
import styles from './TraitsCreation.css';
import Add from '@react/react-spectrum/Icon/Add';

const SingleSignalTraitsCreation = ({
    createTraitUrl,
    traitsCreationLabelText,
    canCreateTrait,
}) => {
    if (canCreateTrait) {
        return (
            <a className="spectrum-Link" href={createTraitUrl}>
                <div className={styles.singleCreation}>
                    <Add size="S" />
                    <div className={styles.singleCreationlinkText}>{traitsCreationLabelText}</div>
                </div>
            </a>
        );
    }

    return '0 Traits';
};

SingleSignalTraitsCreation.propTypes = {
    createTraitUrl: PropTypes.string,
    traitsCreationLabelText: PropTypes.string,
    canCreateTrait: PropTypes.bool,
};

export default SingleSignalTraitsCreation;
