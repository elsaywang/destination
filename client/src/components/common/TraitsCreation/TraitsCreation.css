:local(.singleCreationlinkText) {
    display: inline-flex;
    font-size: 0.875rem;
    padding-left: 0.15em;
    vertical-align: text-bottom;
}

:local(.singleCreation) {
    cursor: pointer;
    margin-left: -1.05em;
    white-space: nowrap;
}

:local(.multiCreation) {
    display: flex;
    justify-content: flex-end;
}

:local(.message) {
    display: flex;
    padding: 0.5em;
    font-weight: 800;
    white-space: nowrap;
}
