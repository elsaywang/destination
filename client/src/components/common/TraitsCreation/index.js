import TraitsCreation from './TraitsCreation';

export default TraitsCreation;
