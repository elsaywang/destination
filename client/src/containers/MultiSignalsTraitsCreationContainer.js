import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import TraitsCreation from '../components/common/TraitsCreation/';
import { getSelectedResults, getSelectedResultsDataSources } from '../reducers';
import { isMaxSignalSelectionsReached, getSelectedSignalType } from '../reducers/selectedSignals';

export class MultiSignalsTraitsCreationContainer extends Component {
    render() {
        return this.props.selectedSignals.selectedRowIndexes.length ? (
            <TraitsCreation multiCreation {...this.props} />
        ) : null;
    }
}

MultiSignalsTraitsCreationContainer.propTypes = {
    signalType: PropTypes.string,
    selectedSignals: PropTypes.shape({
        selectionMessage: PropTypes.string,
        hasSignalSelectionsTypeWarning: PropTypes.bool,
        selectedRowIndexes: PropTypes.array,
    }),
    selectedDataSources: PropTypes.array,
    selectedResults: PropTypes.array,
    canCreateTraits: PropTypes.bool,
    isMaxSignalSelectionsReached: PropTypes.bool,
};

const mapStateToProps = ({ selectedSignals, results, signalType }) => ({
    selectedSignals,
    signalType: getSelectedSignalType(selectedSignals),
    selectedResults: getSelectedResults({
        selectedSignals,
        results,
    }),
    selectedDataSources: getSelectedResultsDataSources({ selectedSignals, results }),
    isMaxSignalSelectionsReached: isMaxSignalSelectionsReached(selectedSignals),
});

export default connect(mapStateToProps)(MultiSignalsTraitsCreationContainer);
