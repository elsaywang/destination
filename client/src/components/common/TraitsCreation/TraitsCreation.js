import React, { Component } from 'react';
import PropTypes from 'prop-types';
import MultiSignalsTraitsCreation from './MultiSignalsTraitsCreation';
import SingleSignalTraitsCreation from './SingleSignalTraitsCreation';
import { createOnboardedTraitUrl, createRuleBasedTraitUrl } from '../../../utils/urls';
import { stringifySignals } from '../../../utils/stringifySignals';

class TraitsCreation extends Component {
    getCreateTraitURL() {
        return this.shouldCreateOnboardedTrait()
            ? createOnboardedTraitUrl()
            : createRuleBasedTraitUrl();
    }

    getLinkText() {
        return this.props.signalType === 'ONBOARDED'
            ? 'Create Onboarded Trait'
            : 'Create Rule-Based Trait';
    }

    getButtonText() {
        return this.props.selectedResults.length > 1
            ? 'Create Trait From Selected Signals'
            : 'Create Trait From Selected Signal';
    }

    shouldCreateOnboardedTrait() {
        // For single signal creation, check the `signalType` of the signal result.
        // For multi signal creation, check if all selected signals are onboarded and
        // come from the same data source.
        const { signalType, doAllSelectedResultsShareSameDataSource } = this.props;

        return signalType === 'ONBOARDED' || doAllSelectedResultsShareSameDataSource;
    }

    storeSessionAndNavigateToTraits = e => {
        e.preventDefault();

        const { keyValuePairs, multiCreation, selectedResults, selectedDataSourceIds } = this.props;

        const signals = multiCreation ? selectedResults : [{ keyValuePairs }];
        const signalsParams = {
            signals: stringifySignals(signals),
            ...(this.shouldCreateOnboardedTrait() && {
                source: {
                    dataSourceIds: selectedDataSourceIds,
                },
            }),
        };

        sessionStorage.setItem('signalsParams', JSON.stringify(signalsParams));

        window.location.assign(this.getCreateTraitURL());
    };

    render() {
        const {
            multiCreation,
            selectedSignals,
            canCreateTraits,
            isMaxSignalSelectionsReached,
        } = this.props;

        return multiCreation ? (
            <MultiSignalsTraitsCreation
                selectedSignals={selectedSignals}
                storeSessionAndNavigateToTraits={this.storeSessionAndNavigateToTraits}
                isMaxSignalSelectionsReached={isMaxSignalSelectionsReached}
                multiTraitCreationButtonText={this.getButtonText()}
            />
        ) : (
            <SingleSignalTraitsCreation
                traitsCreationLabelText={this.getLinkText()}
                canCreateTraits={canCreateTraits}
                storeSessionAndNavigateToTraits={this.storeSessionAndNavigateToTraits}
            />
        );
    }
}

TraitsCreation.propTypes = {
    signalType: PropTypes.string,
    keyValuePairs: PropTypes.array,
    multiCreation: PropTypes.bool,
    selectedSignals: PropTypes.shape({
        selectionMessage: PropTypes.string,
        hasSignalSelectionsTypeWarning: PropTypes.bool,
        selectedRowIndexes: PropTypes.array,
        hasOnboardedSignalSelectionsWarning: PropTypes.bool,
    }),
    selectedDataSourceIds: PropTypes.array,
    canCreateTraits: PropTypes.bool,
    isMaxSignalSelectionsReached: PropTypes.bool,
};

export default TraitsCreation;
