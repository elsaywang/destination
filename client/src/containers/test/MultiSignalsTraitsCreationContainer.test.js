import React from 'react';
import { shallow } from 'enzyme';
import { MultiSignalsTraitsCreationContainer } from '../MultiSignalsTraitsCreationContainer';
import TraitsCreation from '../../components/common/TraitsCreation';

describe('<MultiSignalsTraitsCreationContainer/> component', () => {
    describe('rendering when there is no signal selected', () => {
        const mockFn = jest.fn();
        const initialprops = {
            selectedSignals: {
                selectionMessage: '',
                hasSignalSelectionsTypeWarning: false,
                selectedRowIndexes: [],
            },
        };
        const wrapper = shallow(<MultiSignalsTraitsCreationContainer {...initialprops} />);

        it('matches snapshot', () => {
            expect(wrapper).toMatchSnapshot();
        });

        it('renders null element', () => {
            expect(wrapper.getElement()).toBe(null);
        });
    });

    describe('rendering when there are signals selected', () => {
        const props = {
            selectedSignals: {
                selectionMessage: '1 Real-time signal selected ',
                hasSignalSelectionsTypeWarning: false,
                selectedRowIndexes: [0],
            },
        };
        const wrapper = shallow(<MultiSignalsTraitsCreationContainer {...props} />);

        it('matches snapshot', () => {
            expect(wrapper).toMatchSnapshot();
        });

        it('renders <TraitsCreation/> component', () => {
            expect(wrapper.find(TraitsCreation).exists()).toBe(true);
        });

        it('<TraitsCreation/> is in `Multi-Signals Traits Creation` mode', () => {
            expect(wrapper.find(TraitsCreation).props().multiCreation).toBe(true);
        });

        it('<TraitsCreation/> passes props `selectedSignals`', () => {
            expect(wrapper.find(TraitsCreation).props().selectedSignals).toMatchObject(
                props.selectedSignals,
            );
        });

        it('<TraitsCreation/> passes props `categoryType`', () => {
            expect(wrapper.find(TraitsCreation).props().selectedSignals).toMatchObject(
                props.selectedSignals,
            );
        });
    });
});
