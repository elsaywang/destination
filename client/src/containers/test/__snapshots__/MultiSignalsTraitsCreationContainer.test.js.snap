// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MultiSignalsTraitsCreationContainer/> component rendering when there are signals selected matches snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <MultiSignalsTraitsCreationContainer
    selectedSignals={
        Object {
            "hasWarning": false,
            "records": Array [
              Object {
                "rowIndex": 0,
                "signalType": "Adobe Analytics",
                "source": Object {
                  "sourceType": "ANALYTICS",
                },
              },
            ],
            "selectionMessage": "1 Real-time signal selected ",
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "multiCreation": true,
      "selectedSignals": Object {
        "hasWarning": false,
        "records": Array [
          Object {
            "rowIndex": 0,
            "signalType": "Adobe Analytics",
            "source": Object {
              "sourceType": "ANALYTICS",
            },
          },
        ],
        "selectionMessage": "1 Real-time signal selected ",
      },
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "multiCreation": true,
        "selectedSignals": Object {
          "hasWarning": false,
          "records": Array [
            Object {
              "rowIndex": 0,
              "signalType": "Adobe Analytics",
              "source": Object {
                "sourceType": "ANALYTICS",
              },
            },
          ],
          "selectionMessage": "1 Real-time signal selected ",
        },
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`<MultiSignalsTraitsCreationContainer/> component rendering when there is no signal selected matches snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <MultiSignalsTraitsCreationContainer
    selectedSignals={
        Object {
            "hasWarning": false,
            "records": Array [],
            "selectionMessage": "",
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): null,
  Symbol(enzyme.__nodes__): Array [
    null,
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
